lib/libxdp/libxdp.c

```diff
@@ -1202,6 +1223,54 @@ struct xdp_program *xdp_program__from_pin(const char *pin_path)
 	return prog;
 }
 
+struct xdp_program *xdp_program__create(struct xdp_program_opts *opts)
+{
+	const char *pin_path, *prog_name, *find_filename, *open_filename;
+	struct bpf_object_open_opts *obj_opts;
+	struct bpf_object *obj;
+	__u32 id;
+	int fd;
+
+	if (!opts || !OPTS_VALID(opts, xdp_program_opts))
+		return ERR_PTR(-EINVAL);
+
+	obj           = OPTS_GET(opts, obj, NULL);
+	obj_opts      = OPTS_GET(opts, opts, NULL);
+	prog_name     = OPTS_GET(opts, prog_name, NULL);
+	find_filename = OPTS_GET(opts, find_filename, NULL);
+	open_filename = OPTS_GET(opts, open_filename, NULL);
+	pin_path      = OPTS_GET(opts, pin_path, NULL);
+	id            = OPTS_GET(opts, id, 0);
+	fd            = OPTS_GET(opts, fd, 0);
+
+	if (obj) { /* prog_name is optional */
+		if (obj_opts || find_filename || open_filename || pin_path || id || fd)
+			return ERR_PTR(-EINVAL);
+		return xdp_program__create_from_obj(obj, NULL, prog_name, true);
+	} else if (find_filename) { /* prog_name, obj_opts is optional */
+		if (obj || open_filename || pin_path || id || fd)
+			return ERR_PTR(-EINVAL);
+		return __xdp_program__find_file(find_filename, NULL, prog_name, obj_opts);
+	} else if (open_filename) { /* prog_name, obj_opts is optional */
+		if (obj || find_filename || pin_path || id || fd)
+			return ERR_PTR(-EINVAL);
+		return __xdp_program__open_file(open_filename, NULL, prog_name, obj_opts);
+	} else if (pin_path) {
+		if (obj || obj_opts || prog_name || find_filename || open_filename || id || fd)
+			return ERR_PTR(-EINVAL);
+		return xdp_program__from_pin(pin_path);
+	} else if (id) {
+		if (obj || obj_opts || prog_name || find_filename || open_filename || pin_path || fd)
+			return ERR_PTR(-EINVAL);
+		return xdp_program__from_id(id);
+	} else if (fd) {
+		if (obj || obj_opts || prog_name || find_filename || open_filename || pin_path || id)
+			return ERR_PTR(-EINVAL);
+		return xdp_program__from_fd(fd);
+	}
+	return ERR_PTR(-EINVAL);
```

Comment:
Maybe we should use this opportunity to get rid of the user-facing ERR_PTR() returns? I.e., return NULL and set errno here instead?

We'll have to keep the old behaviour for the existing helpers, but we could modify them to *also* set errno if there's an error before returning. And make sure libxdp_get_error/libxdp_strerror work in all cases...

Reply:
Should this change be made for direct error (integer) returning helpers as well? Like use introduce and use a libxdp_err, libxdp_ptr_err wrappers like libbpf_err, libbpf_err_ptr?

I think then we also need to wrap libbpf calls with it when returning, so that we set errno regardless of strict mode setting, agreed?

Reply:
One thing that is not clear to me is, suppose that I change libxdp_get_error to be like libbpf_get_error:
```
   209 long libxdp_get_error(const void *ptr)                                                                            
   210 {                                                                                                   
   211         if (!IS_ERR_OR_NULL(ptr))                                                                   
   212                 return 0;                                                                           
   213                                                                                                     
   214         if (IS_ERR(ptr))                                                                            
   215                 errno = -PTR_ERR(ptr);                                                              
   216                                                                                                     
   217         /* If ptr == NULL, then errno should be already set by the failing                          
   218          * API, because libxdp never returns NULL on success and it now always                      
   219          * sets errno on error. So no extra errno handling for ptr == NULL                          
   220          * case.                                                                                    
   221          */                                                                                         
   222         return -errno;                                                                              
   223 }                 
```

Then for APIs that already return NULL (e.g. xdp_program__btf), it would be a problem if user does libxdp_get_error on it. It will end up using a stale errno value.

I think then we need to do an all or nothing switch, perhaps based on libbpf strict mode? i.e. add a libxdp_strict_mode setting?